### ISSUE TITLE
Tips For Consumption In An Alpine-based Container Image?

### ISSUE THREAD

**Issue by nujonathan** [seeraven/gitcache#54]
Firstly, I'd like to echo the previously expressed sentiment that this is a very cool project, and I really appreciate it. I'm using it in a custom container image based on [Atlantis'](https://www.runatlantis.io/) image; gitcache dramatically speeds up Atlantis' downloading of GitHub-sourced Terraform modules. Any speed improvements we get further encourages teams to use Atlantis for a communal, highly visible approach to IaC.

To get this deployed, however, I needed to take a fairly circuitous path. In my image build, I'm currently cloning the source, installing the dependencies into a Python virtual environment, then placing a simple wrapper git script early in the `PATH` that just invokes the Python application and passes along any arguments. This works, but is a bit more complicated than I expected.

I wanted to use one of the distributed Linux binaries, but the [base image](https://github.com/runatlantis/atlantis/pkgs/container/atlantis/281293947?tag=v0.30.0) Atlantis distributes is based on Alpine. As far as I can tell, the gitcache binaries are all glibc-linked, and Alpine ships with musl. I attempted to leverage some readily available compatibility layers via apk ([gcompat](https://pkgs.alpinelinux.org/package/edge/main/x86_64/gcompat) , [libc6-compat](https://pkgs.alpinelinux.org/package/v3.15/main/x86_64/libc6-compat)), but there were issues with finding the linker lib. I also tried installing glibc (including [this](https://github.com/sgerrand/alpine-pkg-glibc) Alpine-ready variant), but in these cases, the gitcache binaries immediately seg faulted. Next, I tried building from source, and although I eventually produced a binary, when used as a git wrapper, it would complain about missing Python dependencies (`_socket` is one example, IIRC).

And so, I arrived at the source-based solution outlined above.

It looks like Atlantis does have Debian-based images that would probably be more amenable here, but I'm keen on using Alpine, since the vulnerability counts discovered by tools like Trivy are a good bit lower, in my experience.

Admittedly, I could do more investigating myself, but I did eventually find something that works for both x86 and ARM architectures. For posterity's sake, do you have any plans to distribute binaries that work in Alpine? And maybe with ARM? Failing that, any tips to help build a usable executable in these conditions?

Many thanks for your time and this awesome tool.

**Comment by seeraven** [comment]
Thank you for the very clear and detailed issue description. As you already pointed out, the problem seems to be the non-glibc part, which means that you can run python scripts but not the binary created by the pyInstaller.
I did a quick search and found https://github.com/six8/pyinstaller-alpine as a promising project. It actually builds pyInstaller inside an Alpine docker image and uses that to generate the binary of the python code. If I've understand it correctly, that could work to generate binaries for Alpine.
So a first attempt will be to try to build a pyInstaller generated binary for Alpine 3.20.3 (that was the referenced version in the [atlantis/Dockerfile](https://github.com/runatlantis/atlantis/blob/main/Dockerfile) on x86. Next step would be to put that into a github action to generate the binary whenever a new release is created.
Unfortunately the arm support on github actions is currently non-existing, but I've read it should support arm runners by end of this year/start of next year. That would be the easiest option. Another option would be to implement some qemu virtualization but I feel that is really slow and brittle.
So I'll try the build first locally and keep you informed how it is going. ;-)

**Comment by nujonathan** [comment]
Thanks for the lead! I tried this out quickly as a multi-stage image build, and was able to make it work. Obviously, a lot more care could be taken to match environments, but this did work as a very quick PoC:
```
FROM six8/pyinstaller-alpine-linux-arm64:alpine-3.6-pyinstaller-v3.4 AS build

RUN mkdir -p /gitcache \
  && apk add --no-cache build-base bash python3-dev \
  && git clone --recurse-submodules https://github.com/seeraven/gitcache.git /gitcache \
  && cd /gitcache \
  && make current-release \
  && mv /gitcache/releases/git* /gitcache/gitcache

FROM ghcr.io/runatlantis/atlantis:v0.30.0

COPY --from=build --chown=atlantis /gitcache/gitcache /home/atlantis/bin/gitcache
ENV PATH=/home/atlantis/bin:$PATH
RUN gitcache > /dev/null 2>&1 \
  && mv ~/bin/gitcache ~/bin/git
```
From the container, cloning this repository:
```
51d915d1462d:~$ git clone https://github.com/seeraven/gitcache.git
2024-11-22 19:51:07 Starting Initial clone of https://github.com/seeraven/gitcache.git into /home/atlantis/.gitcache/mirrors/github.com/seeraven/gitcache.
Cloning into bare repository '/home/atlantis/.gitcache/mirrors/github.com/seeraven/gitcache/git'...
remote: Enumerating objects: 1304, done.
remote: Counting objects: 100% (325/325), done.
remote: Compressing objects: 100% (102/102), done.
remote: Total 1304 (delta 216), reused 301 (delta 208), pack-reused 979 (from 1)
Receiving objects: 100% (1304/1304), 337.43 KiB | 2.64 MiB/s, done.
Resolving deltas: 100% (850/850), done.
2024-11-22 19:51:07 Initial clone of https://github.com/seeraven/gitcache.git into /home/atlantis/.gitcache/mirrors/github.com/seeraven/gitcache was successfully completed within 0.6 seconds.
2024-11-22 19:51:08 Starting LFS fetch of ref master from https://github.com/seeraven/gitcache.git into /home/atlantis/.gitcache/mirrors/github.com/seeraven/gitcache.
fetch: Fetching reference refs/heads/master
2024-11-22 19:51:08 LFS fetch of ref master from https://github.com/seeraven/gitcache.git into /home/atlantis/.gitcache/mirrors/github.com/seeraven/gitcache was successfully completed within 0.0 seconds.
2024-11-22 19:51:08 Starting Clone from mirror /home/atlantis/.gitcache/mirrors/github.com/seeraven/gitcache.
Cloning into 'gitcache'...
done.
2024-11-22 19:51:08 Clone from mirror /home/atlantis/.gitcache/mirrors/github.com/seeraven/gitcache was successfully completed within 0.0 seconds.
2024-11-22 19:51:08 Setting push URL to https://github.com/seeraven/gitcache.git and configure LFS.
```

**Comment by seeraven** [comment]
That is really cool, thanks for testing! I guess I'll integrate the Alpine support into the make4py Makefile system similar to the way Ubuntu is supported. That would then allow to also run the functional tests in that environment.

**Comment by seeraven** [comment]
I was able to generate Alpine based binaries using the github actions. Starting with the latest release [v1.0.19](https://github.com/seeraven/gitcache/releases/tag/v1.0.19) Alpine binaries are provided.
Can you check whether that works for you? I've done the integration tests also on Alpine, but you'll never know... ;-)

**Comment by nujonathan** [comment]
> It seems it is not necessary any more to build pyInstaller on Alpine Linux manually like the six8/pyinstaller-alpine-linux-arm64 image does. Simply installing pyInstaller into a python-alpine container uses already the musl bootloader. I've integrated the support for using Alpine docker containers in make4py and it works with my toy example. The next step is to integrate it into the gitcache repository. I think tomorrow I can start working on the github actions so at least binaries for Alpine on x86_64 are generated. Just to confirm, I'm currently using the latest Alpine version 3.20. Would that work for you or do you require an older version?

Yup, current versions of Alpine are preferable. The only reason I used such an old version was that's the latest ARM-compatible image six8 had on Dockerhub :)

**Comment by nujonathan** [comment]
> I was able to generate Alpine based binaries using the github actions. Starting with the latest release [v1.0.19](https://github.com/seeraven/gitcache/releases/tag/v1.0.19) Alpine binaries are provided. Can you check whether that works for you? I've done the integration tests also on Alpine, but you'll never know... ;-)

I've confirmed this works! Many thanks for the efforts.

**Comment by seeraven** [comment]
Cool, thanks for testing. I'll close the issue then. There is another issue regarding arm support, I guess if that github arm runner nodes are available I'll simply generate everything for arm too.
